cabal-install/Distribution/Client/Dependency.hs

```diff
@@ -587,13 +587,19 @@ interpretPackagesPreference :: Set PackageName
                             -> (PackageName -> PackagePreferences)
 interpretPackagesPreference selected defaultPref prefs =
   \pkgname -> PackagePreferences (versionPref pkgname) (installPref pkgname)
-
   where
     versionPref pkgname =
-      fromMaybe anyVersion (Map.lookup pkgname versionPrefs)
-    versionPrefs = Map.fromList
-      [ (pkgname, pref)
-      | PackageVersionPreference pkgname pref <- prefs ]
+      fromMaybe [anyVersion] (Map.lookup pkgname versionPrefs)
+    versionPrefs = foldr mergeIfMultiplePrefs Map.empty prefs
+    mergeIfMultiplePrefs (PackageVersionPreference pkgname pref) m =
+      Map.insertWith (++) pkgname [pref] m
+    -- Previously the two cases below were simply ignored, and I think for a good
+    -- reason, because it is not possible to specify installed or latest preference
+    -- for a package. What to do?
+    mergeIfMultiplePrefs
```

Comment:
`InstalledPreference` is handled below in `installPrefs`, so you can continue filtering it out here.  `interpretPackagesPreference` needs to handle both constructors because sandboxes add preferences on installed instances.